pythran/pythonic/numpy/interp_core.cpp

```diff
@@ -0,0 +1,256 @@
+        if (guess < len - LIKELY_IN_CACHE_SIZE - 1 &&
+            key < arr[guess + LIKELY_IN_CACHE_SIZE]) {
+          imax = guess + LIKELY_IN_CACHE_SIZE;
+        }
+      }
+    }
+  }
+
+  /* finally, find index by bisection */
+  while (imin < imax) {
+    const npy_intp imid = imin + ((imax - imin) >> 1);
+    if (key >= arr[imid]) {
+      imin = imid + 1;
+    } else {
+      imax = imid;
+    }
+  }
+  return imin - 1;
+}
+//
+//#undef LIKELY_IN_CACHE_SIZE
+//
+// NPY_NO_EXPORT PyObject *
+// arr_interp(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
+//{
+//
+//    PyObject *fp, *xp, *x;
+//    PyObject *left = NULL, *right = NULL;
+//    PyArrayObject *afp = NULL, *axp = NULL, *ax = NULL, *af = NULL;
+//    npy_intp i, lenx, lenxp;
+//    npy_double lval, rval;
+//    const npy_double *dy, *dx, *dz;
+//    npy_double *dres, *slopes = NULL;
+//
+//    static char *kwlist[] = {"x", "xp", "fp", "left", "right", NULL};
+//
+//    NPY_BEGIN_THREADS_DEF;
+//
+//    if (!PyArg_ParseTupleAndKeywords(args, kwdict, "OOO|OO:interp", kwlist,
+//                                     &x, &xp, &fp, &left, &right)) {
+//        return NULL;
+//    }
+//
+//    afp = (PyArrayObject *)PyArray_ContiguousFromAny(fp, NPY_DOUBLE, 1, 1);
+//    if (afp == NULL) {
+//        return NULL;
+//    }
+//    axp = (PyArrayObject *)PyArray_ContiguousFromAny(xp, NPY_DOUBLE, 1, 1);
+//    if (axp == NULL) {
+//        goto fail;
+//    }
+//    ax = (PyArrayObject *)PyArray_ContiguousFromAny(x, NPY_DOUBLE, 0, 0);
+//    if (ax == NULL) {
+//        goto fail;
+//    }
+//    lenxp = PyArray_SIZE(axp);
+//    if (lenxp == 0) {
+//        PyErr_SetString(PyExc_ValueError,
+//                        "array of sample points is empty");
+//        goto fail;
+//    }
+//    if (PyArray_SIZE(afp) != lenxp) {
+//        PyErr_SetString(PyExc_ValueError,
+//                        "fp and xp are not of the same length.");
+//        goto fail;
+//    }
+//
+//    af = (PyArrayObject *)PyArray_SimpleNew(PyArray_NDIM(ax),
+//                                            PyArray_DIMS(ax), NPY_DOUBLE);
+//    if (af == NULL) {
+//        goto fail;
+//    }
+//    lenx = PyArray_SIZE(ax);
+//
+//    dy = (const npy_double *)PyArray_DATA(afp);
+//    dx = (const npy_double *)PyArray_DATA(axp);
+//    dz = (const npy_double *)PyArray_DATA(ax);
+//    dres = (npy_double *)PyArray_DATA(af);
+//    /* Get left and right fill values. */
+//    if ((left == NULL) || (left == Py_None)) {
+//        lval = dy[0];
+//    }
+//    else {
+//        lval = PyFloat_AsDouble(left);
+//        if (error_converting(lval)) {
+//            goto fail;
+//        }
+//    }
+//    if ((right == NULL) || (right == Py_None)) {
+//        rval = dy[lenxp - 1];
+//    }
+//    else {
+//        rval = PyFloat_AsDouble(right);
+//        if (error_converting(rval)) {
+//            goto fail;
+//        }
+//    }
+
+// xp->dx   fp->dy  x -> dz
+
+#define PyArray_malloc malloc
```

Comment:
avoid that and use a std::vector that you resize appropriately, that makes memory management safer